integration/src/main/java/scripts/utils/ControlledAzureVmTestScriptBase.java

```diff
@@ -0,0 +1,232 @@
+
+/**
+ * Base class for all Azure VM-based integration tests.
+ *
+ * <p>Defines common user journey for VM-based integration scenarios. It provides default happy path
+ * implementation for vm creation which includes provisioning of network, disk, public
+ * ip(optionally). This behavior can be changed in sub-classes.</>
+ */
+public abstract class ControlledAzureVmTestScriptBase extends ControlledAzureTestScriptBase {
+  private static final Logger logger =
+      LoggerFactory.getLogger(ControlledAzureVmTestScriptBase.class);
+
+  protected static final int VM_CREATE_WAIT_POLL_INTERVAL_SECONDS = 5;
+
+  protected String createVmJobId;
+  // controls if we want to provision VM with public ip
+  protected boolean vmWithPublicIp;
+  protected Optional<CreatedControlledAzureIp> publicIp;
+
+  @Override
+  protected void doSetup(List<TestUserSpecification> testUsers, WorkspaceApi workspaceApi)
+      throws Exception {
+    super.doSetup(testUsers, workspaceApi);
+    vmWithPublicIp = false;
+    publicIp = Optional.empty();
+  }
+
+  @Override
+  protected void doUserJourney(TestUserSpecification testUser, WorkspaceApi workspaceApi)
+      throws Exception {
+    wsmApiClient = ClientTestUtils.getClientForTestUser(testUser, server);
+    azureApi = new ControlledAzureResourceApi(wsmApiClient);
+
+    // create public ip
+    if (vmWithPublicIp) {
+      CreatedControlledAzureIp ip = createPublicIp(azureApi, suffix);
+      publicIp = Optional.of(ip);
+    }
+
+    // create network
+    CreatedControlledAzureNetwork network = createNetwork(azureApi, suffix);
+
+    // create disk
+    CreatedControlledAzureDisk disk = createDisk(azureApi, suffix);
+
+    // create vm
+    createVmJobId = UUID.randomUUID().toString();
+    CreatedControlledAzureVmResult vmCreateResult =
+        createVm(azureApi, suffix, createVmJobId, disk, network);
+
+    vmCreateResult = waitForVmCreationCompletion(azureApi, createVmJobId, vmCreateResult);
+
+    validateVm(vmCreateResult);
+  }
+
+  protected void validateVm(CreatedControlledAzureVmResult vmCreateResult) {
+    assertEquals(JobReport.StatusEnum.SUCCEEDED, vmCreateResult.getJobReport().getStatus());
+    assertNotNull(vmCreateResult.getAzureVm());
+    logger.info(
+        "Create VM name={} status is {}",
+        vmCreateResult.getAzureVm().getAttributes().getVmName(),
+        vmCreateResult.getJobReport().getStatus().toString());
+    logger.info(
+        "Created VM with following parameters Name={}, Resource Id={}",
+        vmCreateResult.getAzureVm().getAttributes().getVmName(),
+        vmCreateResult.getAzureVm().getMetadata().getResourceId());
+  }
+
+  protected abstract CreatedControlledAzureVmResult createVm(
+      ControlledAzureResourceApi azureApi,
+      String resourceSuffix,
+      String createVmJobId,
+      CreatedControlledAzureDisk disk,
+      CreatedControlledAzureNetwork network)
+      throws ApiException;
+
+  protected CreatedControlledAzureIp createPublicIp(
```

Comment:
Would it make sense to pull these method implementations (createPublicIp, createNetwork, etc.) into a utils class? If we're writing additional Azure tests in the future I'd imagine we'll probably want these bits of functionality, even if we aren't making a VM.

Waiting until we actually have additional uses also seems fine too

Reply:
I think we can wait until we see real use cases for this. Anyway, all these methods should be still in this class since it's base class for all VM scenarios, but they can re-use some default behavior from util class.